### PR TITLE
Update payment methods page

### LINKED ISSUE #9639: UA shop: fixing card issue does not trigger a payment retry
**Steps to reproduce:**
1. Make any purchase using card 4242 4242 4242 4242.

2. Change your card to 4000 000 0000 0341 in `/advantage/payment-methods`

3. Try to update the subscription to increase # of machines.

4. The message that shows up says:
    > Payment method:You need to update your payment methods to ensure there is no interruption to your Ubuntu Advantage subscriptions

    While that is somewhat accurate, there is no mention that the subscription increase failed (other than the number not changing). You can distinguish between a purchase for a subscription renewal (which happens in the background) and a purchase that happened due to user action (e.g., subscription increase) failing by looking at the `isRenewal` field of the pending purchase object (upselling/downselling will have this field set to false).

5. Change your card back to 4242 4242 4242 4242 in `/advantage/payment-methods`.

**What is expected**: the card is changed, the payment issues is resolved automatically after the change (by calling `POST /v1/purchase/{purchaseID}/payment/stripe/{invoiceID}`).

**What happens**: the payment is still stuck. Trying to make a new purchase causes the payment modal to be loading forever (rather than displaying an error saying a payment is stuck).

### PR DESCRIPTION
**Pull request by albertkol**
## Done

Update payment methods page.
If they user has a failed payment they will see the following notification at the top of the page:
image

The button will allow them to retry the payment using the same details already stored.

If the payment failed because of 3ds, when they click retry -> the modal will show up:
- if the 3ds succeeds we show the success message and the page reloads after 2 seconds
- if the 3ds fails error message shows up. The user is offered the chance only one time to redo the 3ds, if it fails they have to change the card

If the payment failed because it needs a new payment method, when they click retry -> they will just retry with the same details:
- if the retry succeeds we show the success message and the page reloads after 2 seconds
- if the retry fails, they can still retry

Updating the card will try the payment too. Similar to the retry button.. the same rules apply.

## QA

List of cards to play around with:
- Good card: `4242424242424242`
- Always will ask for 3DS check: `4000002760003184`
- Good card, but will trigger a `card-declined` : `4000000000000341` (this will pass basic validation and will fail only when attempting the purchase)
- Bad card that will trigger a `card-declined`: `4000000000000002` (this will fail basic validation when trying to update card)
- You feel adventurous and what more cards to try out? https://stripe.com/docs/testing#cards

Triggering a 3DS check:
- Update your card to: `4000002760003184`
- Resize up a product
- It will fail and show you the notification
- Go to `/payment-methods` try and fix your purchase, or keep breaking it with other bad credit card numbers:

Triggering a basic require new payment method error:
- Update your card to: `4000000000000341`
- Resize up a product
- It will fail and show you the notification
- Go to `/payment-methods` try and fix your purchase, or keep breaking it with other bad credit card numbers:

## Issue / Card

Fixes #9639

## Screenshots
image

image

image

### CODE REVIEW

**Comment by anthonydillon** [comment]
@albertkol we have a Vanilla pattern for notifications with actions. Can you use this pattern please? https://vanillaframework.io/docs/patterns/notification#actions

**Comment by anasereijo** [comment]
I will wait for UX review but if we decide to go with what you have please follow @anthonydillon's suggestion

> @albertkol we have a Vanilla pattern for notifications with actions. Can you use this pattern please? https://vanillaframework.io/docs/patterns/notification#actions

**Comment by albertkol** [comment]
@wgx I have made the changes you've asked for and made some QOL changes on top.

Two different error message depending on the situation.
- in a situation where you have no pending purchases and you simply want to update your card, but it fails for some reason (test card (test card: `4000 0000 0000 0002`) you will get:
image
- in a situation where you have a pending purchase but the payment retry fails you get:
image

Clicking the re-try link will reveal a spinner/buffer thingy:
image

Can I please get some green +1 :pleading_face:? @wgx @ClementChaumel @anasereijo

**Comment by wgx** [comment]
Happy to oblige with a +1

**Comment by albertkol** [comment]
@wgx @elioqoshi Do you guys think we need to display a message letting people know that the payment will be reattempted immediately after card update (if a payment is indeed pending)?

We can add a short message/disclaimer, somewhere...

**Comment by anthonydillon** [comment]
Could you remove the "Sorry" from the second error notification, please? Notifications should be informative not apologenic.

File: static/js/src/ua-payment-methods.js [hunk 1]
```diff
@@ -89,15 +104,152 @@ updateButton.addEventListener("click", function () {
...
+
+const handlePaymentMethodErrors = (message = null) => {
+  if (!message) {
+    message = "Sorry, there was an error with the payment.";
+  }
+
+  paymentErrorElement.querySelector(".p-notification__message").innerHTML =
+    "<strong>" +
+    message +
+    "</strong> Check the details and try again. Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.";
```
Comment (ClementChaumel) [inline static/js/src/ua-payment-methods.js:151]:
```suggestion
    `<strong>${message}</strong> Check the details and try again. Contact <a class='p-notification__action' href='https://ubuntu.com/contact-us'>Canonical sales</a> if the problem persists.`;
```

this is equivalent to a f-string in python

File: templates/advantage/payment-methods/index.html [hunk 2]
```diff
@@ -32,6 +32,26 @@ <h1>Ubuntu Advantage for&nbsp;Infrastructure</h1>
...
+          </p>
+        </div>
+
+        <div id="payment-warnings" class="p-notification--caution u-hide">
+          <p class="p-notification__response">
+            <span class="p-notification__message">
+              You have one or more failed payments. Use your saved card to <a class="p-notification__action" id="try-again-button">retry payment</a>. <i id="try-again-spinner" class="p-icon--spinner u-animation--spin is-dark u-hide"></i></span>
+          </p>
+        </div>
+
```
Comment (ClementChaumel) [inline templates/advantage/payment-methods/index.html:54]: Here you could have only one notification and change its class (`--positive`, `--negative`, `--caution`) in the JS. 
(I'm not asking you to change it, it's almost just a matter of taste but just letting you know, I don't know if you've considered)

File: static/js/src/ua-payment-methods.js [hunk 3]
```diff
@@ -89,15 +104,152 @@ updateButton.addEventListener("click", function () {
...
+const handleError = () => {
+  updateButton.classList.remove("is-processing");
+  updateButton.innerHTML = "Update";
+  updateButton.disabled = false;
+  if (cancelButton) cancelButton.disabled = false;
+};
+
+const handlePaymentMethodErrors = (message = null) => {
+  if (!message) {
+    message = "Sorry, there was an error with the payment.";
```
Comment (ClementChaumel) [inline static/js/src/ua-payment-methods.js:145]: Remove the Sorry as Ant suggested

```suggestion
    message = "There was an error with the payment.";
```

**Comment by ClementChaumel** [comment]
When I try a failing card as you described here:
>Triggering a basic require new payment method error:
Update your card to: 4000000000000341
Resize up a product
It will fail and show you the notification
Go to /payment-methods try and fix your purchase, or keep breaking it with other bad credit card numbers:

I get the spinner forever 
image

This is the response I get in the network tab

```
HTTP/1.1 0 invoice payment failed
Server: gunicorn
Date: Thu, 13 May 2021 07:26:42 GMT
Connection: keep-alive
Content-Type: application/json
Content-Length: 1658
Cache-Control: no-store
X-View-Name: webapp.advantage.views.post_stripe_invoice_id
Permissions-Policy: interest-cohort=()
X-Frame-Options: SAMEORIGIN
Vary: Cookie
X-VCS-Revision: 1620842289-9d3f15d
X-Request-Id: 0c552ee77f198144db437f91b4a7b652

{"errors":"{\"code\":\"card_declined\",\"decline_code\":\"generic_decline\",\"doc_url\":\"https://stripe.com/docs/error-codes/card-declined\",\"status\":402,\"message\":\"Your card was declined.\",\"payment_intent\":{\"amount\":0,\"amount_capturable\":0,\"amount_received\":0,\"application\":null,\"application_fee_amount\":0,\"canceled_at\":0,\"cancellation_reason\":\"\",\"capture_method\":\"\",\"charges\":null,\"client_secret\":\"\",\"confirmation_method\":\"\",\"created\":0,\"currency\":\"\",\"customer\":null,\"description\":\"\",\"invoice\":null,\"last_payment_error\":null,\"livemode\":false,\"id\":\"pi_1IqYvKCzjFajHovd9yuMumP2\",\"metadata\":null,\"next_action\":null,\"on_behalf_of\":null,\"payment_method\":null,\"payment_method_options\":null,\"payment_method_types\":null,\"receipt_email\":\"\",\"review\":null,\"setup_future_usage\":\"\",\"shipping\":{\"address\":null,\"carrier\":\"\",\"name\":\"\",\"phone\":\"\",\"tracking_number\":\"\"},\"source\":null,\"statement_descriptor\":\"\",\"statement_descriptor_suffix\":\"\",\"status\":\"\",\"transfer_data\":null,\"transfer_group\":\"\"},\"payment_method\":{\"afterpay_clearpay\":null,\"alipay\":null,\"au_becs_debit\":null,\"bacs_debit\":null,\"bancontact\":null,\"billing_details\":null,\"card\":null,\"card_present\":null,\"created\":0,\"customer\":null,\"eps\":null,\"fpx\":null,\"giropay\":null,\"grabpay\":null,\"id\":\"pm_1IqYv2CzjFajHovdNSU5Vynz\",\"ideal\":null,\"interac_present\":null,\"livemode\":false,\"metadata\":null,\"object\":\"\",\"oxxo\":null,\"p24\":null,\"sepa_debit\":null,\"sofort\":null,\"type\":\"\"},\"request_id\":\"req_gqM6gCy8Cw10BN\",\"type\":\"card_error\"}"}
```

**Comment by wgx** [comment]
@albertkol I think it's implicit, no need for another line of text.

File: templates/advantage/payment-methods/index.html (hunk 2 above)
Comment (albertkol) [reply templates/advantage/payment-methods/index.html:54]: I could make the `--positive` and the `--negative` the same div.

.. but not sure how I would merge the `--negative` and the `--caution` to be the same div as they can show up both at the same time.

But is the merge worth it? I feel like having 3 separate elements is sooo much easier, as I just need to reveal them. Once you only have one element you have to perform more complex operations to update the message and colour of the same div.

Comment (ClementChaumel) [reply templates/advantage/payment-methods/index.html:54]: Oh, I didn't realise the caution and negative could show at the same time. I don't think it's worth it, to change it if you prefer it this way it's was just a suggestion!

**Comment by albertkol** [comment]
Having a pending purchase view:
image

Resolving the pending purchase errors out:
image

Pending payment buffering:
image

Resolving pending purchase succeeds:
image

@anasereijo I guess there is not much design to review..

**Comment by albertkol** [comment]
@ClementChaumel I was unable to reproduce the problem. Maybe I fixed it involuntarily with the other changes, although it should not have changed anything about the problem you've described.

Could you please try again and let me know if it's still happening.

File: static/js/src/ua-payment-methods.js (hunk 3 above)
Comment (albertkol) [reply static/js/src/ua-payment-methods.js]: Changed!

**Comment by ClementChaumel** [comment]
> @ClementChaumel I was unable to reproduce the problem. Maybe I fixed it involuntarily with the other changes, although it should not have changed anything about the problem you've described.
> 
> Could you please try again and let me know if it's still happening.

I'm sorry I can still see the issue. Let's jump on a call tomorrow?

**Comment by albertkol** [comment]
@ClementChaumel back with you for review. The problem was actually much bigger than I could have expected. The application was not setting the status_code of there request correctly.

On localhost the web server was able to parse the json response even if the status_code was missing, but the demo was not so it would get stuck.

**Comment by ClementChaumel** [comment]
> @ClementChaumel back with you for review. The problem was actually much bigger than I could have expected. The application was not setting the status_code of there request correctly.
> 
> On localhost the web server was able to parse the json response even if the status_code was missing, but the demo was not so it would get stuck.

Ooh sneaky :bug: !
It works correctly now, thanks!

**Review by ClementChaumel — APPROVED** [review]

**Pull request merged**